___

<p style="text-align: center;"><img src="https://docs.google.com/uc?id=1lY0Uj5R04yMY3-ZppPWxqCr5pvBLYPnV" class="img-fluid" alt="CLRSWY"></p>

___

# WELCOME!

Welcome to "RFM Customer Segmentation & Cohort Analysis Project". This is the first project of the Capstone Project Series, which consists of 3 different project that contain different scenarios.

This is a project which you will learn what is RFM? And how to apply RFM Analysis and Customer Segmentation using K-Means Clustering. Also you will improve your Data Cleaning, Data Visualization and Exploratory Data Analysis capabilities. On the other hand you will create Cohort and Conduct Cohort Analysis. 

Before diving into the project, please take a look at the determines and project structure.

- **NOTE:** This Project assumes that you already know the basics of coding in Python and are familiar with the theory behind K-Means Clustering.



# #Determines

Using the [Online Retail dataset](https://archive.ics.uci.edu/ml/datasets/Online+Retail) from the UCI Machine Learning Repository for exploratory data analysis, ***Customer Segmentation***, ***RFM Analysis***, ***K-Means Clustering*** and ***Cohort Analysis***.

This is a transnational data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail. The company mainly sells unique all-occasion gifts. Many customers of the company are wholesalers.

Feature Information:

**InvoiceNo**: Invoice number. *Nominal*, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation. 
<br>
**StockCode**: Product (item) code. *Nominal*, a 5-digit integral number uniquely assigned to each distinct product.
<br>
**Description**: Product (item) name. *Nominal*. 
<br>
**Quantity**: The quantities of each product (item) per transaction. *Numeric*.
<br>
**InvoiceDate**: Invoice Date and time. *Numeric*, the day and time when each transaction was generated.
<br>
**UnitPrice**: Unit price. *Numeric*, Product price per unit in sterling.
<br>
**CustomerID**: Customer number. *Nominal*, a 5-digit integral number uniquely assigned to each customer.
<br>
**Country**: Country name. *Nominal*, the name of the country where each customer resides.


---


First of all, to observe the structure of the data and missing values, you can use exploratory data analysis and data visualization techniques.

You must do descriptive analysis. Because you must understand the relationship of the features to each other and clear the noise and missing values in the data. After that, the data set will be ready for RFM analysis.

Before starting the RFM Analysis, you will be asked to do some analysis regarding the distribution of *Orders*, *Customers* and *Countries*. These analyzes will help the company develop its sales policies and contribute to the correct use of resources.

You will notice that the UK not only has the most sales revenue, but also the most customers. So you will continue to analyze only UK transactions in the next RFM Analysis, Customer Segmentation and K-Means Clustering topics.

Next, you will begin RFM Analysis, a customer segmentation technique based on customers' past purchasing behavior. 

By using RFM Analysis, you can enable companies to develop different approaches to different customer segments so that they can get to know their customers better, observe trends better, and increase customer retention and sales revenues.

You will calculate the Recency, Frequency and Monetary values of the customers in the RFM Analysis you will make using the data consisting of UK transactions. Ultimately, you have to create an RFM table containing these values.

In the Customer Segmentation section, you will create an RFM Segmentation Table where you segment your customers by using the RFM table. For example, you can label the best customer as "Big Spenders" and the lost customer as "Lost Customer".

We will segment the customers ourselves based on their recency, frequency, and monetary values. But can an **unsupervised learning** model do this better for us? You will use the K-Means algorithm to find the answer to this question. Then you will compare the classification made by the algorithm with the classification you have made yourself.

Before applying K-Means Clustering, you should do data pre-processing. In this context, it will be useful to examine feature correlations and distributions. In addition, the data you apply for K-Means should be normalized.

On the other hand, you should inform the K-means algorithm about the number of clusters it will predict. You will also try the *** Elbow method *** and *** Silhouette Analysis *** to find the optimum number of clusters.

After the above operations, you will have made cluster estimation with K-Means. You should visualize the cluster distribution by using a scatter plot. You can observe the properties of the resulting clusters with the help of the boxplot. Thus you will be able to tag clusters and interpret results.

Finally, you will do Cohort Analysis with the data you used at the beginning, regardless of the analysis you have done before. Cohort analysis is a subset of behavioral analytics that takes the user data and breaks them into related groups for analysis. This analysis can further be used to do customer segmentation and track metrics like retention, churn, and lifetime value.


# #Project Structures

- Data Cleaning & Exploratory Data Analysis
- RFM Analysis
- Customer Segmentation
- Applying K-Means Clustering
- Create Cohort and Conduct Cohort Analysis

# #Tasks

#### 1. Data Cleaning & Exploratory Data Analysis

- Import Modules, Load Data & Data Review
- Follow the Steps Below

    *i. Take a look at relationships between InvoiceNo, Quantity and UnitPrice columns.*
    
    *ii. What does the letter "C" in the invoiceno column mean?*
    
    *iii. Handling Missing Values*
    
    *iv. Clean the Data from the Noise and Missing Values*
    
    *v. Explore the Orders*
    
    *vi. Explore Customers by Country*
    
    *vii. Explore the UK Market*
    
#### 2. RFM Analysis

- Follow the steps below

   *i. Import Libraries*
   
   *ii. Review "df_uk" DataFrame (the df_uk what you create at the end of the Task 1)*
   
   *iii. Calculate Recency*
   
   *iv. Calculate Frequency*
   
   *v. Calculate Monetary Values*
   
   *vi. Create RFM Table*

#### 3. Customer Segmentation with RFM Scores
- Calculate RFM Scoring

    *i. Creating the RFM Segmentation Table*
 
- Plot RFM Segments

#### 4. Applying K-Means Clustering
- Data Pre-Processing and Exploring

    *i. Define and Plot Feature Correlations*
 
    *ii. Visualize Feature Distributions*
 
    *iii. Data Normalization*

- K-Means Implementation

    *i. Define Optimal Cluster Number (K) by using "Elbow Method" and "Silhouette Analysis"*
 
    *ii. Visualize the Clusters*
 
    *iii. Assign the label*
 
    *iv. Conclusion*
 
#### 5. Create Cohort and Conduct Cohort Analysis
- Future Engineering

    *i. Extract the Month of the Purchase*
 
    *ii. Calculating time offset in Months i.e. Cohort Index*
 
- Create 1st Cohort: User Number & Retention Rate 

    *i. Pivot Cohort and Cohort Retention*
 
    *ii. Visualize analysis of cohort 1 using seaborn and matplotlib*

- Create 2nd Cohort: Average Quantity Sold 

    *i. Pivot Cohort and Cohort Retention*
 
    *ii. Visualize analysis of cohort 2 using seaborn and matplotlib*

- Create 3rd Cohort: Average Sales

    *i. Pivot Cohort and Cohort Retention*
 
    *ii. Visualize analysis of cohort 3 using seaborn and matplotlib*
    
- **Note: There may be sub-tasks associated with each task, you will see them in order during the course of the work.**


# 1. Data Cleaning & Exploratory Data Analysis

## Import Modules, Load Data & Data Review

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns 
import matplotlib.pyplot as plt

import warnings                      
warnings.filterwarnings("ignore")
warnings.warn("this will not show")

%matplotlib inline
# %matplotlib notebook   bu notebook zorlar ancak interaktif olarak imkan sağlar.

plt.rcParams["figure.figsize"] = (10,6)   # tüm grafikleri bu ölcüye sabitledik.
# plt.rcParams['figure.dpi'] = 100

sns.set_style("whitegrid")             # çizime özel yapılabilsede gridleri beyaz yaptı.
pd.set_option('display.float_format', lambda x: '%.3f' % x)   # tüm floatları virgülden sonra 3 karakter verir.

pd.options.display.max_rows = 100
pd.options.display.max_columns = 100

In [2]:
df = pd.read_excel("Online Retail.xlsx", 
                   dtype={"InvoiceNo": str, "StockCode":str,"CustomerID": str},
                   parse_dates=['InvoiceDate'])
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0  536365.0    85123A   WHITE HANGING HEART T-LIGHT HOLDER     6.000   
1  536365.0   71053.0                  WHITE METAL LANTERN     6.000   
2  536365.0    84406B       CREAM CUPID HEARTS COAT HANGER     8.000   
3  536365.0    84029G  KNITTED UNION FLAG HOT WATER BOTTLE     6.000   
4  536365.0    84029E       RED WOOLLY HOTTIE WHITE HEART.     6.000   

          InvoiceDate  UnitPrice CustomerID         Country  
0 2010-12-01 08:26:00      2.550    17850.0  United Kingdom  
1 2010-12-01 08:26:00      3.390    17850.0  United Kingdom  
2 2010-12-01 08:26:00      2.750    17850.0  United Kingdom  
3 2010-12-01 08:26:00      3.390    17850.0  United Kingdom  
4 2010-12-01 08:26:00      3.390    17850.0  United Kingdom

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  float64       
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  object        
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), object(5)
memory usage: 33.1+ MB


In [4]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [5]:
df.describe(include= "all").T

count unique                                 top    freq  \
InvoiceNo       541909  25900                            573585.0    1114   
StockCode       541909   4070                              85123A    2313   
Description     540455   4223  WHITE HANGING HEART T-LIGHT HOLDER    2369   
Quantity    541909.000    NaN                                 NaN     NaN   
InvoiceDate     541909  23260                 2011-10-31 14:41:00    1114   
UnitPrice   541909.000    NaN                                 NaN     NaN   
CustomerID      406829   4372                             17841.0    7983   
Country         541909     38                      United Kingdom  495478   

                          first                last  mean     std        min  \
InvoiceNo                   NaT                 NaT   NaN     NaN        NaN   
StockCode                   NaT                 NaT   NaN     NaN        NaN   
Description                 NaT                 NaT   NaN     NaN        NaN   
Quantity                    NaT                 NaT 9.552 218.081 -80995.000   
InvoiceDate 2010-12-01 08:26:00 2011-12-09 12:50:00   NaN     NaN        NaN   
UnitPrice                   NaT                 NaT 4.611  96.760 -11062.060   
CustomerID                  NaT                 NaT   NaN     NaN        NaN   
Country                     NaT                 NaT   NaN     NaN        NaN   

              25%   50%    75%       max  
InvoiceNo     NaN   NaN    NaN       NaN  
StockCode     NaN   NaN    NaN       NaN  
Description   NaN   NaN    NaN       NaN  
Quantity    1.000 3.000 10.000 80995.000  
InvoiceDate   NaN   NaN    NaN       NaN  
UnitPrice   1.250 2.080  4.130 38970.000  
CustomerID    NaN   NaN    NaN       NaN  
Country       NaN   NaN    NaN       NaN

In [11]:
df.describe().T

count  mean     std        min   25%   50%    75%       max
Quantity  541909.000 9.552 218.081 -80995.000 1.000 3.000 10.000 80995.000
UnitPrice 541909.000 4.611  96.760 -11062.060 1.250 2.080  4.130 38970.000

In [6]:
df.skew()

Quantity      -0.264
UnitPrice    186.507
CustomerID     0.030
dtype: float64

In [7]:
df.kurtosis()

Quantity     119769.160
UnitPrice     59005.719
CustomerID       -1.180
dtype: float64

In [8]:
df.shape

(541909, 8)

In [9]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [10]:
df.isnull().sum()*100/df.shape[0]

InvoiceNo      0.000
StockCode      0.000
Description    0.268
Quantity       0.000
InvoiceDate    0.000
UnitPrice      0.000
CustomerID    24.927
Country        0.000
dtype: float64

### i. Take a look at relationships between InvoiceNo, Quantity and UnitPrice columns.

In [13]:
df[['InvoiceNo', 'Quantity', 'UnitPrice']].head()

InvoiceNo  Quantity  UnitPrice
0  536365.0     6.000      2.550
1  536365.0     6.000      3.390
2  536365.0     8.000      2.750
3  536365.0     6.000      3.390
4  536365.0     6.000      3.390

In [16]:
df[['InvoiceNo', 'Quantity', 'UnitPrice']].describe()

Quantity  UnitPrice
count 541909.000 541909.000
mean       9.552      4.611
std      218.081     96.760
min   -80995.000 -11062.060
25%        1.000      1.250
50%        3.000      2.080
75%       10.000      4.130
max    80995.000  38970.000

In [15]:
df['InvoiceNo'].nunique()

25900

In [14]:
df.groupby('InvoiceNo').count()[['Quantity', 'UnitPrice']]

Quantity  UnitPrice
InvoiceNo                     
536365.0          7          7
536366.0          2          2
536367.0         12         12
536368.0          4          4
536369.0          1          1
...             ...        ...
C581484           1          1
C581490           2          2
C581499           1          1
C581568           1          1
C581569           2          2

[25900 rows x 2 columns]

We see that there are negative values in the Quantity and UnitPrice columns. These are possibly canceled and returned orders. Let's check it out.

In [17]:
df.loc[df.InvoiceNo.str.contains("C"),["InvoiceNo","Description"]]

InvoiceNo                       Description
141      C536379                          Discount
154      C536383   SET OF 3 COLOURED  FLYING DUCKS
235      C536391    PLASTERS IN TIN CIRCUS PARADE 
236      C536391  PACK OF 12 PINK PAISLEY TISSUES 
237      C536391  PACK OF 12 BLUE PAISLEY TISSUES 
...          ...                               ...
540449   C581490   ZINC T-LIGHT HOLDER STARS SMALL
541541   C581499                            Manual
541715   C581568        VICTORIAN SEWING BOX LARGE
541716   C581569  HANGING HEART JAR T-LIGHT HOLDER
541717   C581569     36 PENCILS TUBE RED RETROSPOT

[9288 rows x 2 columns]

In [22]:
df[~(df['InvoiceNo'].str.contains('C')) & (df['Quantity'] < 0)]['CustomerID'].value_counts()

Series([], Name: CustomerID, dtype: int64)

In [23]:
df.groupby(['StockCode']).nunique()['CustomerID']

StockCode
10002.0         40
10080.0         19
10120.0         25
10123C           3
10123G           0
                ..
gift_0001_20     0
gift_0001_30     0
gift_0001_40     0
gift_0001_50     0
m                0
Name: CustomerID, Length: 4070, dtype: int64

### ii. What does the letter "C" in the InvoiceNo column mean?

If the invoice number starts with the letter "C", it means the order was cancelled. Or those who abandon their order.

When we filter canceled orders by Quantity> 0 or filter non-canceled orders by Quantity <0 nothing returns, this confirms that negative values mean the order was canceled. So lets find out how many orders were cancelled?

#### 9288 or about 36% of the orders were cancelled. Looking deeper into why these orders were cancelled may prevent future cancellations. Now let's find out what a negative UnitPrice means.


### iii. Handling Missing Values

Since the customer ID's are missing, lets assume these orders were not made by the customers already in the data set because those customers already have ID's. 

We also don't want to assign these orders to those customers because this would alter the insights we draw from the data. 


### iv. Clean the Data from the Noise and Missing Values

### v. Explore the Orders


1. Find the unique number of InvoiceNo  per customer

2. What's the average number of unqiue items per order or per customer?

3. Let's see how this compares to the number of unique products per customer.

### vi. Explore Customers by Country

1. What's the total revenue per country?

2. Visualize number of customer per country

3. Visualize total cost per country

#### The UK not only has the most sales revenue, but also the most customers. Since the majority of this data set contains orders from the UK, we can explore the UK market further by finding out what products the customers buy together and any other buying behaviors to improve our sales and targeting strategy.

### vii. Explore the UK Market


1. Create df_uk DataFrame

2. What are the most popular products that are bought in the UK?

### We will continue analyzing the UK transactions with customer segmentation.

# 2. RFM Analysis

In the age of the internet and e-commerce, companies that do not expand their businesses online or utilize digital tools to reach their customers will run into issues like scalability and a lack of digital precsence. An important marketing strategy e-commerce businesses use for analyzing and predicting customer value is customer segmentation. Customer data is used to sort customers into group based on their behaviors and preferences.

**[RFM](https://www.putler.com/rfm-analysis/) (Recency, Frequency, Monetary) Analysis** is a customer segmentation technique for analyzing customer value based on past buying behavior. RFM analysis was first used by the direct mail industry more than four decades ago, yet it is still an effective way to optimize your marketing.
<br>
<br>
Our goal in this Notebook is to cluster the customers in our data set to:
 - Recognize who are our most valuable customers
 - Increase revenue
 - Increase customer retention
 - Learn more about the trends and behaviors of our customers
 - Define customers that are at risk

We will start with **RFM Analysis** and then compliment our findings with predictive analysis using **K-Means Clustering Algorithms.**

- RECENCY (R): Time since last purchase
- FREQUENCY (F): Total number of purchases
- MONETARY VALUE (M): Total monetary value




Benefits of RFM Analysis

- Increased customer retention
- Increased response rate
- Increased conversion rate
- Increased revenue

RFM Analysis answers the following questions:
 - Who are our best customers?
 - Who has the potential to be converted into more profitable customers?
 - Which customers do we need to retain?
 - Which group of customers is most likely to respond to our marketing campaign?
 

### i. Import Libraries

### ii. Review df_uk DataFrame

### iii. Recency: Days since last purchase
To calculate the recency values, follow these steps in order:

1. To calculate recency, we need to choose a date as a point of reference to evaluate how many days ago was the customer's last purchase.
2. Create a new column called Date which contains the invoice date without the timestamp
3. Group by CustomerID and check the last date of purchase
4. Calculate the days since last purchase
5. Drop Last_Purchase_Date since we don't need it anymore
6. Plot RFM distributions

1. Choose a date as a point of reference to evaluate how many days ago was the customer's last purchase.

2. Create a new column called Date which contains the invoice date without the timestamp

3. Group by CustomerID and check the last date of purchase

4. Calculate the days since last purchase

5. Drop Last_Purchase_Date since we don't need it anymore

6. Plot RFM distributions

### iv. Frequency: Number of purchases

To calculate how many times a customer purchased something, we need to count how many invoices each customer has. To calculate the frequency values, follow these steps in order:

1. Make a copy of df_uk and drop duplicates

2. Calculate the frequency of purchases

3. Plot RFM distributions

### v. Monetary: Total amount of money spent

The monetary value is calculated by adding together the cost of the customers' purchases.


1. Calculate sum total cost by customers and named "Monetary"

2. Plot RFM distributions

### vi. Create RFM Table
Merge the recency, frequency and motetary dataframes

# 3. Customer Segmentation with RFM Scores

Businesses have this ever-lasting urge to understand their customers. The better you understand the customer, the better you serve them, and the higher the financial gain you receive from that customer. Since the dawn of trade, this process of understanding customers for a strategic gain has been there practiced and this task is known majorly as [Customer Segmentation](https://clevertap.com/blog/rfm-analysis/).
Well as the name suggests, Customer Segmentation could segment customers according to their precise needs. Some of the common ways of segmenting customers are based on their Recency-Frequency-Monatory values, their demographics like gender, region, country, etc, and some of their business-crafted scores. You will use Recency-Frequency-Monatory values for this case.

In this section, you will create an RFM Segmentation Table where you segment your customers by using the RFM table. For example, you can label the best customer as "Big Spenders" and the lost customer as "Lost Customer".

## Calculate RFM Scoring

The simplest way to create customer segments from an RFM model is by using **Quartiles**. We will assign a score from 1 to 4 to each category (Recency, Frequency, and Monetary) with 4 being the highest/best value. The final RFM score is calculated by combining all RFM values. For Customer Segmentation, you will use the df_rfm data set resulting from the RFM analysis.
<br>
<br>
**Note**: Data can be assigned into more groups for better granularity, but we will use 4 in this case.

1. Divide the df_rfm into quarters

### i. Creating the RFM Segmentation Table


1. Create two functions, one for Recency and one for Frequency and Monetary. For Recency, customers in the first quarter should be scored as 4, this represents the highest Recency value. Conversely, for Frequency and Monetary, customers in the last quarter should be scored as 4, representing the highest Frequency and Monetary values.

2. Score customers from 1 to 4 by applying the functions you have created. Also create separate score column for each value. 

3. Now that scored each customer, you'll combine the scores for segmentation.

4. Define rfm_level function that tags customers by using RFM_Scrores and Create a new variable RFM_Level

5. Calculate average values for each RFM_Level, and return a size of each segment 

## Plot RFM Segments

1. Create your plot and resize it.

Using customer segmentation categories found [here](http://www.blastam.com/blog/rfm-analysis-boosts-sales) we can formulate different marketing strategies and approaches for customer engagement for each type of customer.

Note: The author in the article scores 1 as the highest and 4 as the lowest

2. How many customers do we have in each segment?

# 3. Applying K-Means Clustering

Now that we have our customers segmented into 6 different categories, we can gain further insight into customer behavior by using predictive models in conjuction with out RFM model.
Possible algorithms include **Logistic Regression**, **K-means Clustering**, and **K-nearest Neighbor**. We will go with [K-Means](https://towardsdatascience.com/understanding-k-means-clustering-in-machine-learning-6a6e67336aa1) since we already have our distinct groups determined. K-means has also been widely used for market segmentation and has the advantage of being simple to implement.

## Data Pre-Processing and Exploring

### i. Define and Plot Feature Correlations

Create Heatmap and evaluate the results 

### ii. Visualize Feature Distributions

To get a better understanding of the dataset, you can costruct a scatter matrix of each of the three features in the RFM data.

### iii. Data Normalization

1. You can use the logarithm method to normalize the values in a column.

2. Plot normalized data with scatter matrix or pairplot. Also evaluate results.

## K-Means Implementation

For k-means, you have to set k to the number of clusters you want, but figuring out how many clusters is not obvious from the beginning. We will try different cluster numbers and check their [silhouette coefficient](http://scikit-learn.org/stable/auto_examples/cluster/plot_kmeans_silhouette_analysis.html). The silhouette coefficient for a data point measures how similar it is to its assigned cluster from -1 (dissimilar) to 1 (similar). 
<br>
<br>
**Note**: K-means is sensitive to initializations because they are critical to qualifty of optima found. Thus, we will use smart initialization called "Elbow Method".

### i. Define the Optimal Number of Clusters

[The Elbow Method](https://en.wikipedia.org/wiki/Elbow_method_(clustering) 

[Silhouette Coefficient](http://scikit-learn.org/stable/auto_examples/cluster/plot_kmeans_silhouette_analysis.html)

### ii. Model Fitting

Fit the K-Means Algorithm with the optimal number of clusters you decided and save the model to disk.

### iii. Visualize the Clusters

1. Create a scatter plot and select cluster centers

2. Visualize Cluster Id vs Recency, Cluster Id vs Frequency and Cluster Id vs Monetary using Box plot. Also evaluate the results. 

### iv. Assign the Label

**Conclusion**

- Cluster 0 : The first cluster belongs to the "Best Customers" segment which we saw earlier as they purchase recently (R=4), frequent buyers (F=4), and spent the most (M=4)

- Cluster 1 : Second cluster can be interpreted as passer customers as their last purchase is long ago (R<=1),purchased very few (F>=2 & F < 4) and spent little (M>=4 & M < 4).Company has to come up with new strategies to make them permanent members. Low value customers
- Cluster 2 : The third cluster is more related to the "Almost Lost" segment as they Haven’t purchased for some time(R=1), but used to purchase frequently and spent a lot.

### v. Conclusion

Discuss your final results. Compare your own labels from the Customer Segmentation with the labels found by K-Means.

How we want to continue this analysis depends on how the business plans to use the results and the level of granularity the business stakeholders want to see in the clusters. We can also ask what range of customer behavior from high to low value customers are the stakeholders interested in exploring. From those answers, various methods of clustering can be used and applied on RFM variable or directly on the transaction data set.

**Annotation:**

Limitations of K-means clustering:

1. There is no assurance that it will lead to the global best solution.
2. Can't deal with different shapes(not circular) and consider one point's probability of belonging to more than one cluster.

These disadvantages of K-means show that for many datasets (especially low-dimensional datasets), it may not perform as well as you might hope.

# 5. Create Cohort & Conduct Cohort Analysis
[Cohort Analysis](https://medium.com/swlh/cohort-analysis-using-python-and-pandas-d2a60f4d0a4d) is specifically useful in analyzing user growth patterns for products. In terms of a product, a cohort can be a group of people with the same sign-up date, the same usage starts month/date, or the same traffic source.
Cohort analysis is an analytics method by which these groups can be tracked over time for finding key insights. This analysis can further be used to do customer segmentation and track metrics like retention, churn, and lifetime value.

For e-commerce organizations, cohort analysis is a unique opportunity to find out which clients are the most valuable to their business. by performing Cohort analysis you can get the following answers to the following questions:

- How much effective was a marketing campaign held in a particular time period?
- Did the strategy employ to improve the conversion rates of Customers worked?
- Should I focus more on retention rather than acquiring new customers?
- Are my customer nurturing strategies effective?
- Which marketing channels bring me the best results?
- Is there a seasonality pattern in Customer behavior?
- Along with various performance measures/metrics for your organization.

Since we will be performing Cohort Analysis based on transaction records of customers, the columns we will be dealing with mainly:
- Invoice Data
- CustomerID
- Price
- Quantity

The following steps will performed to generate the Cohort Chart of Retention Rate:
- Month Extraction from InvioceDate column
- Assigning Cohort to Each Transaction
- Assigning Cohort Index to each transaction
- Calculating number of unique customers in each Group of (ChortDate,Index)
- Creating Cohort Table for Retention Rate
- Creating the Cohort Chart using the Cohort Table

The Detailed information about each step is given below:

## Future Engineering

### i. Extract the Month of the Purchase
First we will create a function, which takes any date and returns the formatted date with day value as 1st of the same month and Year.

Now we will use the function created above to convert all the invoice dates into respective month date format.

### ii. Calculating time offset in Months i.e. Cohort Index:
Calculating time offset for each transaction will allows us to report the metrics for each cohort in a comparable fashion.
First, you will create 4 variables that capture the integer value of years, months for Invoice and Cohort Date using the get_date_int() function which you'll create it below.

You will use this function to extract the integer values for Invoice as well as Cohort Date in 3 seperate series for each of the two columns

Use the variables created above to calcualte the difference in days and store them in cohort Index column.

## Create 1st Cohort: User number & Retention Rate

### i. Pivot Cohort and Cohort Retention

### ii. Visualize analysis of cohort 1 using seaborn and matplotlib modules

## Create the 2nd Cohort: Average Quantity Sold

### i. Pivot Cohort and Cohort Retention

### ii. Visualize analysis of cohort 2 using seaborn and matplotlib modules

## Create the 3rd Cohort: Average Sales


### i. Pivot Cohort and Cohort Retention

### ii. Visualize analysis of cohort 3 using seaborn and matplotlib modules

For e-commerce organisations, cohort analysis is a unique opportunity to find out which clients are the most valuable to their business. by performing Cohort analysis you can get answers to following questions:

- How much effective was a marketing campaign held in a particular time period?
- Did the strategy employed to improve the conversion rates of Customers worked?
- Should I focus more on retention rather than acquiring new customers?
- Are my customer nurturing strategies effective?
- Which marketing channels bring me the best results?
- Is there a seasoanlity pattern in Customer behahiour?

___

<p style="text-align: center;"><img src="https://docs.google.com/uc?id=1lY0Uj5R04yMY3-ZppPWxqCr5pvBLYPnV" class="img-fluid" alt="CLRSWY"></p>

___